# Analysis of Amazon Product Data

After construction of the graph (product-id and link to cross-buys), the following topics will be investigated:
* correlation between in-degrees and salesrank
* and some further smart questions... influence of number of reviews?, are there different sub-graphs like produkt categories?
* 
* 
* 
Data source: https://snap.stanford.edu/data/amazon-meta.html

Data import

In [3]:
# Import graphframes (from Spark-Packages)

from graphframes import *
from pyspark.sql.functions import monotonically_increasing_id, desc


# Get the product data (already pre-processed data in table format)
products = spark.read.csv("/FileStore/tables/ProductData_small.csv", sep=";", header=True)

# The the cross-buy data (similar products)
links = spark.read.csv("/FileStore/tables/LinkData_small.csv", sep=";", header=True)

How many products are we dealing with here?

In [5]:
products.count()

Out[ 10 ]: 10611

quick check of the data formats

In [7]:
display(products.take(3))

Id ASIN Group Salesrank Reviews Title 21 0790747324 DVD 795 140 The Time Machine 84 B000063W82 Video 780 14 The Best of Schoolhouse Rock! - 30th Anniversary Edition 252 B0000262WI Music 581 82 The Köln Concert

And which and how many groups are there?

In [9]:
display(products.select("Group").distinct())

Group Video Toy DVD Sports Baby Product Video Games Book Music Software

In [10]:
products.select("Group").distinct().count()

Out[ 13 ]: 9

And how many products per category?

In [12]:
display(products.groupBy('Group').count().sort('count', ascending=False))

Group count Video 2933 Music 2855 Book 2535 DVD 2276 Software 5 Toy 4 Sports 1 Baby Product 1 Video Games 1

In [13]:
display(links.take(3))

src dest 0790747324 B00007JMD8 0790747324 B00004RF9B 0790747324 B00005JKFR

The graph will be constructed based on the product data. Each product is identified through its unique *ASIN* (Amazon Standard Identification Number) code

GraphFrame library required *ASIN* to be renamed to *id*, but be careful, there is already an "Id" in the DataFrame

In [15]:
# rename "Id" to "Id-nu" to avoid any confusion (nu = not used)
# rename "ASIN" to "id"
ProductVertices = products.withColumnRenamed("Id", "Id-nu").distinct()
ProductVertices = ProductVertices.withColumnRenamed("ASIN", "id").distinct()

# check if done correctly
display(ProductVertices.take(3))

Id-nu id Group Salesrank Reviews Title 13260 B000002TIR Music 2004 16 French Kiss 13683 0140281916 Book 2877 16 Bargaining for Advantage : Negotiation Strategies for Reasonable People 30773 B000001EYO Music 3579 31 The Big Lebowski: Original Motion Picture Soundtrack

How does the Edge table look like?

In [17]:
display(links.take(1))

src dest 0790747324 B00007JMD8

Although not always strictly required, it's good practice to have an ID column for the tables. 

Since the product links does not have one yet, create and additional ID:

In [19]:
# 2. Create Edges
links = links.withColumn("link-id", monotonically_increasing_id())

display(links.take(3))

src dest link-id 0790747324 B00007JMD8 0 0790747324 B00004RF9B 1 0790747324 B00005JKFR 2

Salesrank and number of reviews are read as string. Change type to int for calculations.

In [21]:
# change type of features "Salesrank" and "Reviews" to int
ProductVertices = ProductVertices.withColumn('Salesrank', ProductVertices.Salesrank.cast('int'))
ProductVertices = ProductVertices.withColumn('Reviews', ProductVertices.Reviews.cast('int'))
display(ProductVertices.take(1))

Id-nu id Group Salesrank Reviews Title 13260 B000002TIR Music 2004 16 French Kiss

Rename links to similar products ar *src* and *dst*

In [23]:
# rename destination "dest" to "dst"
links = links.withColumnRenamed('dest', 'dst')


Now comes the moment of truth! Let's see if the graph can be constructed!

In [25]:
# Build the GraphFrame based on products and links

ProductGraph = GraphFrame(ProductVertices, links)

# print a sample of the vertices and edges
print("Products: %d" % ProductGraph.vertices.count())
print
print("SAMPLE PRODUCT DATA:")
print(ProductGraph.vertices.take(1))
print

print("Links: %d" % ProductGraph.edges.count())
print
print("SAMPLE LINK DATA:")
print(ProductGraph.edges.take(1))
print

Products: 10611
SAMPLE PRODUCT DATA:
[Row(Id-nu='13260', id='B000002TIR', Group='Music', Salesrank=2004, Reviews=16, Title='French Kiss')]
Links: 21560
SAMPLE LINK DATA:
[Row(src='0790747324', dst='B00007JMD8', link-id=0)]
 Out[ 21 ]: <function print>

The Graph is constructed. Do the fancy analyses!

1) assess the number of in-degress per product and compare it to the salesrank

Plot chunk 24 in the following way:

you can look at the distribution of the inDegree-values by clicking on the icon to the right of 'Raw table' and then chosing a histogram, values = 'inDegree' and  30 bins

We can see that most of the inDegree values are between 0 and 2

In [29]:
display(ProductGraph.inDegrees.sort('inDegree', ascending=False))
#inDegRank = ProductGraph.inDegrees.sort('inDegree', ascending=False)

id inDegree B00008LDNZ 78 B000068DBC 37 6304765266 36 B00005ATZT 34 B00001QEE7 33 6305368171 33 B00003CXA2 30 B000059HB7 29 630522577X 29 B000022TSH 27 B00003CX95 27 6305696071 27 0684801523 27 B000000XB9 26 0316769487 24 B00005QFDX 23 B000065DWB 23 B00003CXCQ 22 B0000026NF 22 B00000F14B 22 B00001QEE6 21 B00005Y1XY 21 B00005V3Z4 21 0446310786 20 B00004RF97 20 B00005LKHZ 20 B00005LC4A 20 B00005KARK 20 0790731479 20 B00008975Z 20 B00005YUPQ 20 6305280770 19 0694003611 19 0790729628 19 B00008CMR4 19 B00005NTNW 19 B00003CXE7 19 B000067J1P 19 0140177396 19 B00001MXXH 19 B00003CX8Y 18 0066620996 18 0399501487 18 B00000K31Q 18 0399226907 18 B00006RCO1 18 B00005JKOI 18 0800141660 18 B00005Q79A 18 0792839749 18 6305168857 18 0805047905 18 B00006FDCP 18 0792165055 17 0449214923 17 B00005NQJJ 17 B00003JRCQ 17 B00003CXBU 17 6305144168 17 B00000JS62 17 B000002ADT 16 B000023VTP 16 B00004U104 16 0783225725 16 0064401847 16 B00005JKDR 16 B00004XPPG 16 B00004KD24 16 B00005PJ8V 16 6305216088 16 B000055ZF6 16 B000053VBK 16 B00003CXC7 16 0790729377 16 0440237688 16 B000055Y0X 16 0783117523 15 1568322267 15 0553210092 15 B00000DHZJ 15 B00005S6K8 15 6304696566 15 1568328478 15 B00002CGGH 15 6305212112 15 B000035Z1Z 15 0792833287 15 B00004VVNB 15 6303144535 15 0440498058 15 B00003CXD3 15 B00003CXAT 15 B000056WQX 14 B0000058HV 14 0451526341 14 B0000062XS 14 B00005R87R 14 6305428476 14 B00006JMRE 14 B00005AAA9 14 B00005ASGF 14 B00005YUPN 14 B00005N5S3 14 B00007LVCM 14 B00004TDTO 14 B00004RF9B 14 B00000ILEC 14 B00005B6JZ 14 1568328605 14 B00005QJEF 14 B00008ZZ7B 14 B00007J88T 14 0790749815 14 630529142X 14 B00005Q4CS 14 B00003CX9E 14 6303467814 14 0064430227 14 0060254920 14 B00006AL1E 14 B00005MLVA 13 0780621611 13 B000065UFD 13 B00008NRI8 13 B00004Y62W 13 B00005JKHN 13 B00005O3VC 13 B00007GZR1 13 B0000542C9 13 B00006IR69 13 0783240171 13 B00000I21X 13 B00005PJ8S 13 B00007E2F5 13 B00005LZT9 13 B000002KAR 13 6304907729 13 B0000033WI 13 0689835604 13 B000087F7D 13 B00001QEE9 13 B000002G3Y 13 B00007JMD8 13 6304711891 13 0071381392 13 0783227884 13 6305428042 13 0312422156 13 079284615X 13 0609610570 13 B00001QEE8 13 6304711913 13 B000002TTP 13 0375810455 13 B0000648WV 13 0738920762 13 B00005221M 13 B00005JSHW 13 078322592X 13 B00000G3AZ 13 B000000W5P 13 B00005MP58 13 B00005PJ8O 12 0345342968 12 B00003CXDC 12 0800141741 12 B000063K1U 12 0790740060 12 B00004TJKK 12 189176702X 12 B000065V3P 12 B00006JSS8 12 0060256656 12 B000067FP3 12 B00003RQNJ 12 B000001AMZ 12 630227608X 12 6304696612 12 0738920770 12 6304698690 12 B00000K3CQ 12 B000002KHJ 12 B00005JKFR 12 B00008438T 12 B00004RF9D 12 0440414806 12 B00003CXCZ 12 B00003CWL6 12 B00003CXDD 12 B00004R99Q 12 B000002UAO 12 0394800168 12 B00005N7YZ 12 B00002RAPT 12 B00005JKG1 12 B000094J7Z 12 0767803434 12 B00000IBME 12 B0000648WU 12 B0000541AK 12 B000065U1M 12 0790731037 12 B00000IQW3 12 0767830555 12 6304698526 12 B00005LOKQ 12 0783233493 12 0767812131 11 B000006OS2 11 B000056BP4 11 6305907145 11 6304696531 11 B00005J6UP 11 6305428352 11 B00008LDO2 11 6304907613 11 B00005R09Z 11 0792843592 11 B00003CXC5 11 B00006DEF9 11 B00005QFE1 11 B00003CXBK 11 B000050FEN 11 0792164962 11 B00005ALMH 11 B00001U0E1 11 1568328443 11 B000059XUT 11 B00000K3TN 11 B00000FC5H 11 B00004WG2F 11 B000002U8G 11 630469623X 11 0790731525 11 B00005Y72C 11 043935806X 11 0439064864 11 6305161917 11 B00003CX9G 11 6305511055 11 B00005RH91 11 B00005B234 11 B00005ATQB 11 B000002TSS 11 B00003CXA4 11 B00004R99J 11 6304698801 11 6305310343 11 B000002OO0 11 0780622553 11 B00000G3PA 11 B000065U1N 11 B00005N5S5 11 0440413281 10 B00004R9A4 10 0440136482 10 B000002KD9 10 B00003OST5 10 B00000K3CR 10 B0000714BZ 10 0792846133 10 B00006AALQ 10 0790732203 10 B000055Z4M 10 0790739739 10 B000002GVS 10 B000001FDV 10 B00004R993 10 B000063UR0 10 0767818172 10 0446677450 10 0783226055 10 B00008K76V 10 B000002LLR 10 1573441449 10 B00007ELEL 10 6304153066 10 B00000IPS5 10 B000002UB3 10 0486280861 10 B00004Z4WX 10 0783240287 10 B00005JLEM 10 6304698682 10 B00000K0E6 1

In [30]:
maxNumInDeg = ProductGraph.inDegrees.groupBy().max('inDegree')
display(maxNumInDeg)

maxNumInDegValue = str(maxNumInDeg.collect()[0]['max(inDegree)'])
#print(maxNumInDegValue)

maxInDegreeProduct = ProductGraph.inDegrees.filter('inDegree = {}'.format(maxNumInDegValue))
display(maxInDegreeProduct)


id inDegree B00008LDNZ 78

In [31]:
# join back to get product info based on max inDegree
ProductMaxInDegree = ProductGraph.vertices.join(maxInDegreeProduct, 'id')
display(ProductMaxInDegree)




id Id-nu Group Salesrank Reviews Title inDegree B00008LDNZ 548091 DVD 110 83 Laura 78

Plot the following chunk in this way:
Plot this as a bar chart with 'Plot Options':
 - Keys: InDegree
 - Values: Salesrank
 
Unfortunately you can't chose to have a logarithmic y-axis, but you do see that at least for low in-degree values the sales rank is rather bad.

Alternative: scatterplot inDegree vs. salesrank and grouping by product group (Keys=Group)

In [33]:
# repeat exercise from above but with entire list
selection = ProductGraph.inDegrees.sort('inDegree', ascending=False)
#display(selection)

ProductSortInDeg = ProductGraph.vertices.join(selection, 'id').sort('inDegree', ascending=False)
display(ProductSortInDeg)

id Id-nu Group Salesrank Reviews Title inDegree B00008LDNZ 548091 DVD 110 83 Laura 78 B000068DBC 546689 DVD 109 633 Pulp Fiction (Collector's Edition) 37 6304765266 97579 DVD 351 152 While You Were Sleeping 36 B00005ATZT 547040 DVD 1020 140 The Fugitive (Special Edition) 34 6305368171 276371 DVD 236 500 You've Got Mail 33 B00001QEE7 104775 DVD 49 160 The Little Mermaid (Limited Issue) 33 B00003CXA2 547272 DVD 185 530 Forrest Gump 30 B000059HB7 4388 DVD 1877 79 Rio Bravo 29 630522577X 548182 DVD 253 180 My Fair Lady 29 0684801523 199628 Book 956 934 The Great Gatsby 27 B000022TSH 544390 DVD 1108 154 Chinatown 27 6305696071 545472 DVD 540 193 Pretty Woman (10th Anniversary Edition) 27 B00003CX95 166635 DVD 233 745 Braveheart 27 B000000XB9 327081 Music 87 163 """Creedence Clearwater Revival - Chronicle, Vol. 1: The 20 Greatest Hits""" 26 0316769487 98756 Book 60 2568 The Catcher in the Rye 24 B00005QFDX 220569 DVD 330 115 Sesame Street - The Best of Elmo 23 B000065DWB 46130 Music 417 31 Foreigner - Complete Greatest Hits 23 B0000026NF 3499 Music 153 180 Journey - Greatest Hits 22 B00003CXCQ 402685 DVD 679 247 Snow White and the Seven Dwarfs (Disney Special Platinum Edition) 22 B00000F14B 216263 DVD 351 123 The Music Man (Special Edition) 22 B00005V3Z4 546221 DVD 176 848 Donnie Darko 21 B00005Y1XY 429734 Music 333 47 Cars - Complete Greatest Hits 21 B00001QEE6 444195 DVD 904 68 Lady and the Tramp (Limited Issue) 21 0446310786 537519 Book 111 1414 To Kill a Mockingbird 20 B00005LKHZ 164067 DVD 343 236 The NeverEnding Story 20 B00005KARK 453361 DVD 353 140 Dumbo (60th Anniversary Edition) 20 B00005LC4A 545385 DVD 1260 158 Charade 20 B00008975Z 270909 DVD 695 407 Reservoir Dogs 20 B00004RF97 96590 DVD 792 104 The Philadelphia Story 20 B00005YUPQ 28804 DVD 389 129 Baby Einstein - Baby Van Gogh - World of Colors 20 0790731479 138550 DVD 633 126 Beetlejuice 20 6305280770 546760 DVD 1915 79 The King and I 19 0790729628 545235 DVD 392 812 Blade Runner (The Director's Cut) 19 B00008CMR4 547333 DVD 1798 38 Big Jake 19 0694003611 502086 Book 156 339 Goodnight Moon (Board Book) 19 B00001MXXH 548318 DVD 652 295 Ferris Bueller's Day Off 19 B00003CXE7 520382 DVD 1702 1513 Gladiator 19 B000067J1P 545271 DVD 1331 357 The Sound of Music (Single Disc Full Screen Edition) 19 0140177396 341570 Book 126 1000 Of Mice and Men (Penguin Great Books of the 20th Century) 19 B00005NTNW 548266 DVD 768 78 The Outlaw Josey Wales 19 B00005JKOI 137583 DVD 2582 75 Pretty in Pink 18 B00006FDCP 59459 DVD 1582 224 One Flew Over the Cuckoo's Nest (Two-Disc Special Edition) 18 0800141660 546277 DVD 443 250 Sense and Sensibility 18 0399226907 180888 Book 279 164 The Very Hungry Caterpillar board book 18 B00003CX8Y 191667 DVD 439 354 Beauty and the Beast (Disney Special Platinum Edition) 18 B00005Q79A 191370 DVD 1169 166 What's Eating Gilbert Grape 18 B00000K31Q 18824 DVD 1379 387 10 Things I Hate About You 18 0399501487 35512 Book 143 1101 Lord of the Flies 18 0066620996 154855 Book 29 363 Good to Great: Why Some Companies Make the Leap... and Others Don't 18 B00006RCO1 544551 DVD 979 162 All About Eve (Special Edition) 18 0792839749 12389 DVD 384 125 Chitty Chitty Bang Bang (Full Screen Edition) 18 0805047905 502784 Book 171 172 Brown Bear, Brown Bear, What Do You See? 18 6305168857 152642 DVD 154 234 Young Frankenstein (Special Edition) 18 0792165055 544654 DVD 1993 163 Romeo & Juliet 17 B00003CXBU 543078 DVD 1278 119 On the Waterfront (Special Edition) 17 6305144168 544508 DVD 972 276 Dead Poets Society 17 0449214923 482263 Book 215 260 Think and Grow Rich 17 B00003JRCQ 424377 DVD 462 204 Persuasion 17 B00000JS62 547782 DVD 55 380 The Wizard of Oz 17 B00005NQJJ 469947 Music 257 26 Doobie Brothers - Greatest Hits 17 B00004KD24 462983 Music 771 25 The Lovin' Spoonful - Greatest Hits 16 B000023VTP 547943 DVD 505 329 Notting Hill (Collector's Edition) 16 6305216088 190895 DVD 586 355 Good Will Hunting (Miramax Collector's Series) 16 B00004U104 100749 DVD 2126 83 If These 

Plot inDegree vs. Salesrank for Books only
* plot as scatter plot with LOESS smoother

In [35]:
# try to differentiate by product group
ProductSortInDeg = ProductGraph.vertices.join(selection, 'id').sort('inDegree', ascending=False).filter("Group = 'Book'")
display(ProductSortInDeg)

id Id-nu Group Salesrank Reviews Title inDegree 0684801523 199628 Book 956 934 The Great Gatsby 27 0316769487 98756 Book 60 2568 The Catcher in the Rye 24 0446310786 537519 Book 111 1414 To Kill a Mockingbird 20 0694003611 502086 Book 156 339 Goodnight Moon (Board Book) 19 0140177396 341570 Book 126 1000 Of Mice and Men (Penguin Great Books of the 20th Century) 19 0805047905 502784 Book 171 172 Brown Bear, Brown Bear, What Do You See? 18 0399226907 180888 Book 279 164 The Very Hungry Caterpillar board book 18 0066620996 154855 Book 29 363 Good to Great: Why Some Companies Make the Leap... and Others Don't 18 0399501487 35512 Book 143 1101 Lord of the Flies 18 0449214923 482263 Book 215 260 Think and Grow Rich 17 0064401847 42948 Book 1258 655 Bridge to Terabithia 16 0440237688 188765 Book 428 2617 The Giver 16 0440498058 291216 Book 414 872 A Wrinkle in Time 15 0553210092 317053 Book 368 348 The Scarlet Letter 15 0064430227 206514 Book 1032 76 Harold and the Purple Crayon 50th Anniversary Edition (Purple Crayon Books) 14 0060254920 268369 Book 148 211 Where the Wild Things Are 14 0451526341 53175 Book 275 990 Animal Farm 14 0071381392 18173 Book 239 355 The No-Cry Sleep Solution: Gentle Ways to Help Your Baby Sleep Through the Night 13 0312422156 425735 Book 133 517 Middlesex: A Novel 13 0375810455 220181 Book 209 20 Too Big for Diapers (Too Big Board Books) 13 0609610570 429788 Book 143 157 Execution: The Discipline of Getting Things Done 13 0689835604 335281 Book 250 90 Where Is Baby's Belly Button? 13 0394800168 261029 Book 895 148 Green Eggs and Ham (I Can Read It All by Myself Beginner Books) 12 0060256656 239578 Book 360 394 The Giving Tree 12 0345342968 29543 Book 81 1069 Fahrenheit 451 12 189176702X 532007 Book 1725 8 "How to be the Leader of the Pack...And have Your Dog Love You For It. (""How to"" booklets from Dog's Best Friend)" 12 0440414806 528497 Book 761 2922 Holes (Yearling Newbery) 12 0439064864 216201 Book 445 2422 Harry Potter and the Chamber of Secrets (Book 2) 11 043935806X 429348 Book 110 5545 Harry Potter and the Order of the Phoenix (Book 5) 11 0345446798 240297 Book 2251 77 The Other End of the Leash 10 0486280861 296016 Book 122 13 Easy Spanish Phrase Book: Over 770 Basic Phrases for Everyday Use 10 0142001740 337971 Book 26 996 The Secret Life of Bees 10 1573441449 436250 Book 467 32 The Ultimate Guide to Cunnilingus: How to Go Down on a Woman and Give Her Exquisite Pleasure (Ultimate Everything!!!) 10 0671723650 53885 Book 64 377 How to Win Friends & Influence People 10 0440136482 532042 Book 161 351 Holy Blood, Holy Grail 10 0446677450 162283 Book 37 1637 Rich Dad, Poor Dad: What the Rich Teach Their Kids About Money--That the Poor and Middle Class Do Not! 10 0440413281 40399 Book 930 350 Bud, Not Buddy (Newbery Medal Winner, 2000) 10 0486291138 515301 Book 290 35 1001 Most Useful Spanish Words (Beginners' Guides) 10 0787960756 33690 Book 89 61 The Five Dysfunctions of a Team: A Leadership Fable 10 076360013X 33247 Book 295 209 Guess How Much I Love You (Guess How Much I Love You) 10 0671708635 28211 Book 405 583 The 7 Habits of Highly Effective People 9 0142000663 211438 Book 201 517 The Grapes of Wrath : (Centennial Edition) 9 0684852861 357720 Book 215 189 First, Break All the Rules: What the World's Greatest Managers Do Differently 9 0689711816 179199 Book 1414 251 From the Mixed-Up Files of Mrs. Basil E. Frankweiler 9 0440439884 105186 Book 1457 400 Island of the Blue Dolphins 9 0440227534 492127 Book 695 548 Number the Stars (Laurel Leaf Books) 9 0789205386 62274 Book 353 158 The Expectant Father: Facts, Tips and Advice for Dads-To-Be 9 0446677469 234402 Book 384 155 Rich Dad's Guide to Investing: What the Rich Invest in, That the Poor and the Middle Class Do Not! 9 0439136350 404634 Book 356 2396 Harry Potter and the Prisoner of Azkaban (Book 3) 9 0590353403 526761 Book 170 5039 Harry Potter and the Sorcerer's Stone (Book 1) 8 0399144463 406114 Book 126 1279 Who Moved My Cheese? An Amazing 

Plot the following chunk in this way:
Plot this as a bar chart with 'Plot Options':
 - Keys: OutDegree
 - Values: Salesrank

In [37]:
# repeat exercise from above for outDegree
selection = ProductGraph.outDegrees.sort('outDegree', ascending=True)
#display(selection)

ProductSortOutDeg = ProductGraph.vertices.join(selection, 'id').sort('OutDegree', ascending=True)
display(ProductSortOutDeg)

id Id-nu Group Salesrank Reviews Title outDegree 0312421850 6709 Book 4030 124 A Yellow Raft in Blue Water: A Novel 1 630310374X 308746 Video 357 14 The Egg and I 1 0787968331 96240 Book 710 40 The Leadership Challenge, 3rd Edition 1 B00007GZYC 542859 DVD 1200 41 The Crossing 1 B00003CXUM 168136 DVD 3787 100 In the Mood for Love - Criterion Collection 1 6300270270 435731 Video 3587 98 The Bad Seed 1 031021923X 271744 Book 2784 125 Jesus I Never Knew, The 1 1573623806 215324 Video 4386 86 Marilyn Manson - Dead to the World 1 B000002H38 520466 Music 2161 92 Rapture 1 B000000JO0 353679 Music 2144 58 13 Songs 1 B00004VW4Q 71081 Music 3675 21 Partridge Family Album 1 B00005QATI 526297 Video 2431 53 Uprising 1 0865713499 160391 Book 4641 5 A Volcano in My Tummy: Helping Children to Handle Anger : A Resource Book for Parents, Caregivers and Teachers 1 031242227X 104729 Book 327 464 Running with Scissors: A Memoir 1 B00006AUHB 377405 DVD 2416 2 As Time Goes By - Complete Series 3 1 0789447010 193413 Book 1829 45 The Visual Dictionary of Star Wars, Episode I - The Phantom Menace 1 6302689201 42573 Video 4453 12 The Little Engine That Could 1 B00000J5W1 444176 Music 2927 20 Legendary John Denver 1 0316545562 375072 Book 875 141 A World Lit Only by Fire : The Medieval Mind and the Renaissance - Portrait of an Age 1 B000065894 344052 Music 4315 169 Alive or Just Breathing 1 B00000272D 531309 Music 4796 52 Social Distortion 1 B0000034ON 4339 Music 4556 5 Let's Dance, Vol. 5: Modern Collection (Michelle) 1 6302078024 293126 Video 1087 20 The Man Who Never Was 1 B000003TAP 269006 Music 1160 250 August and Everything After 1 0062506048 500318 Book 2193 24 Facing Love Addiction: Giving Yourself the Power to Change the Way You Love --The Love Connection to Codependence 1 6300183289 1263 Video 4411 14 The Lonely Lady 1 B000063K2U 82575 DVD 3358 31 The Saddle Club - Adventures at Pine Hollow 1 B000063K5Q 107085 Video 547 14 Tae-Bo Focus Series - Abs & Glutes 1 0156711427 463953 Book 2866 84 A Passage to India 1 096063889X 299940 Book 1740 9 A Course in Miracles Combined Volume 2nd Ed: I. Text II. Workbook for Students III. Manual for Teachers 1 B000055ZGY 362036 Video 905 36 Twitch Of The Death Nerve (a.k.a. Bay Of Blood) 1 B00005Y1YA 453662 Music 2024 22 Six Feet Under 1 0446610399 349687 Book 1045 234 The Rescue 1 0679406832 160904 Book 612 50 Gift from the Sea : 50th Anniversary Edition 1 B00004Y7CU 215112 Video 4609 1 The Best of Mission: Impossible Vol.10 1 0553348973 274294 Book 2952 141 Still Life with Woodpecker 1 B00006SFKH 201011 DVD 3737 4 Yoga Zone Ultimate Collection 1 B000046S2F 547392 DVD 4860 7 The Twilight Zone: Vol. 8 1 0898154081 504311 Book 1989 27 Thinkertoys (A Handbook of Business Creativity) 1 B000002WXX 57264 Music 2968 19 Hot Tuna 1 6305971935 546264 DVD 1852 26 The Philadelphia Experiment 1 B000000OM9 103862 Music 2874 29 Dreamgirls (1982 Original Broadway Cast) 1 0805366245 320405 Book 772 84 Biology, Sixth Edition 1 B00005N8AL 97777 DVD 4209 43 Bellydance Fitness for Beginners - Arms, Abs, Hips, Buns & Thighs 1 B00000273M 331847 Music 727 59 The Bangles - Greatest Hits 1 0786808381 193399 Book 2360 10 Baby Einstein: Babies (Baby Einstein) 1 6300185850 107428 Video 634 22 Night Mother 1 B00006B1HI 143162 DVD 1296 145 9/11 - The Filmmakers' Commemorative Edition 1 0780621638 441668 Video 575 26 The Winter Guest 1 B00004WG34 543535 Video 553 1 Story of Gospel Music: The Power in the Voice 1 0768356881 267232 Video 3870 16 Land of the Pharaohs 1 0395248302 273616 Book 4090 39 The Stars : A New Way to See Them 1 B00007BH5C 149317 Music 1230 188 The Lord of the Rings: The Two Towers 1 0446673803 179523 Book 1822 52 What Your Doctor May Not Tell You About(TM): Premenopause, Balance Your Hormones and Your Life from Thirty to Fifty 1 0425179613 311906 Book 84 582 It's Not About the Bike: My Journey Back to Life 1 0471430676 122840 Book 3451 20 The Wellness Revolution : How to Make a Fortune in the Next Trillion Dollar Industry

outDegree is rather boring, since most of the products get assigned 5 products by default that are presumingly similar, as you can see here:

In [39]:
display(ProductSortOutDeg.groupBy("outDegree").count())

outDegree count 1 2379 2 2163 3 1840 4 1430 5 723

plot the output of chunk 31 in the following way:
chose a bar chart with options:
  - keys: Reviews
  - values: Salesrank
  
and apply to all rows.
We can see that the less reviews a product has, the less likely it is to be a top selling product (the causal orientation might also be inverse here :) )

In [41]:
# relation of number of reviews and sales rank:
display(ProductGraph.vertices.select("Reviews", "Salesrank").sort("Reviews", ascending = False))


Reviews Salesrank 5545 110 5539 2135 5539 420 5039 170 5039 1533 5034 746 4924 661 4924 404 4924 327 3839 3221 3520 1616 3520 3494 3518 1676 3518 187 3051 3468 3049 19 2925 2264 2925 487 2922 761 2617 428 2568 3735 2568 60 2562 2872 2562 28 2562 511 2562 1599 2422 3807 2422 445 2419 1227 2396 2569 2396 356 2295 142 2280 551 2167 313 2097 85 2097 565 2091 17 2072 1388 2026 662 2026 3649 1985 214 1983 648 1936 2199 1927 1647 1927 654 1894 2165 1869 1073 1867 4574 1855 1983 1826 280 1826 4207 1826 1576 1826 337 1818 420 1814 1425 1814 4836 1814 4748 1811 757 1811 267 1726 2985 1697 2059 1695 160 1695 1378 1695 1233 1688 3162 1637 3459 1637 37 1635 471 1627 455 1552 31 1551 898 1540 1790 1513 1702 1513 770 1513 4615 1501 4273 1460 486 1458 676 1432 1817 1423 401 1414 111 1414 3396 1414 3089 1410 4220 1382 3159 1377 3085 1377 42 1318 709 1315 46 1300 4000 1300 493 1293 1567 1291 599 1285 2218 1279 126 1278 539 1278 3505 1278 1038 1261 1662 1244 782 1223 2679 1223 133 1200 851 1199 123 1196 969 1191 975 1186 1667 1185 4107 1182 3634 1173 4231 1173 85 1169 3014 1169 1491 1156 131 1156 723 1143 4979 1143 1587 1142 339 1108 238 1103 4761 1101 4372 1101 143 1101 4831 1101 1592 1100 1302 1094 1934 1094 1372 1094 2592 1091 571 1069 3681 1069 81 1060 712 1059 622 1051 2714 1048 1734 1046 3853 1030 1991 1019 3708 1016 1234 1015 1596 1011 428 1006 1578 1004 1179 1002 1842 1000 126 996 2121 996 26 994 4287 993 722 990 227 990 762 990 275 989 1301 988 4679 982 491 974 539 974 1401 972 69 969 489 961 1018 958 1320 957 33 955 862 955 3786 952 894 934 956 933 4429 931 427 927 273 926 1403 926 2206 925 298 925 184 924 661 921 3346 918 3334 903 2326 881 105 880 836 873 1476 872 414 869 712 869 128 863 3956 861 4226 861 293 855 151 855 2986 855 419 849 156 848 176 847 754 840 1321 837 2471 835 782 829 1012 827 174 826 919 824 1097 822 526 819 723 817 3538 817 2136 816 32 814 579 812 392 812 4690 810 533 810 2959 809 1075 808 232 805 562 805 424 803 42 803 3057 801 1704 799 1395 796 102 794 395 793 804 792 1929 792 796 786 3503 786 47 785 176 785 3981 784 1963 781 242 780 3147 780 1592 777 668 772 4751 771 342 771 1303 771 182 770 3429 770 3280 765 3019 761 1305 760 1790 758 2502 757 3049 755 998 755 4122 755 1240 753 767 751 114 746 1324 746 619 746 21 745 3092 745 233 745 305 744 497 734 1072 732 195 730 552 730 3870 729 669 728 135 722 4453 721 2471 720 1388 720 1605 718 1215 712 75 712 492 712 970 709 2025 709 996 707 61 707 1356 705 194 703 1825 702 1606 702 3734 701 265 701 1572 697 2744 697 1874 693 1836 692 2816 692 295 692 1414 688 429 688 1716 685 4396 683 89 682 107 682 2955 681 4579 681 157 680 1060 679 3367 677 224 677 269 677 965 677 4492 676 185 676 2857 676 3331 676 4493 675 3417 675 4968 674 628 670 125 670 853 655 1258 653 1750 652 3147 650 2877 648 2403 647 223 647 2917 647 1572 647 1344 642 596 639 2295 639 2697 636 2743 636 111 636 482 634 1179 633 109 633 713 633 1023 631 3319 631 134 630 581 630 1875 628 1992 628 2494 626 412 626 4113 624 4030 622 2488 620 4213 619 2627 619 661 617 4550 617 2937 616 517 612 1973 612 1433 612 4262 612 575 611 1906 610 3049 610 3334 609 1229 609 2903 609 377 609 80 608 454 606 1154 603 1348 602 3300 602 4061 601 3237 600 174 600 333 600 4832 598 4275 598 1032 597 2825 597 1739 595 660 595 3111 594 1524 593 4272 593 585 592 1466 591 1944 590 2505 589 1229 587 67 586 4191 586 38 586 2621 586 1042 584 2217 583 4930 583 3961 583 405 583 2191 582 3884 582 84 582 1049 579 217 578 986 578 1280 576 2171 574 3807 572 273 572 483 572 1043 571 685 571 2162 569 708 567 1599 567 113 567 570 567 4820 566 1126 565 2452 564 470 563 4410 562 3320 562 1532 562 268 562 529 561 2457 561 4694 559 539 557 255 557 300 556 3715 556 786 555 432 555 2986 554 1100 553 2345 552 3727 551 2171 549 4023 548 178 548 695 546 1252 546 413 545 2439 544 4229 544 1962 543 764 543 1683 543 2405 541 117 541 515 539 1165 538 2299 537 234 537 988 537 855 535 2942 534 1073 531 3160 531 681 530 4769 530 185 529 149 529 1636 528 2979 527 1

trying to find products which have an edge to another product which in turn has an edge to the first product:

In [43]:
ProductsWithBalancedSimilarity = ProductGraph.find("(v1)-[e1]->(v2); (v2)-[e2]->(v1)")
display(ProductsWithBalancedSimilarity)

v1 e1 v2 e2 List(21, 0790747324, DVD, 795, 140, The Time Machine) List(0790747324, B00007JMD8, 0) List(114681, B00007JMD8, DVD, 1076, 104, Journey to the Center of the Earth) List(B00007JMD8, 0790747324, 4581) List(84, B000063W82, Video, 780, 14, The Best of Schoolhouse Rock! - 30th Anniversary Edition) List(B000063W82, 156949407X, 4) List(239294, 156949407X, Video, 1223, 24, Schoolhouse Rock! - Grammar Rock) List(156949407X, B000063W82, 9363) List(296, 0385504209, Book, 19, 3049, The Da Vinci Code) List(0385504209, 0671027360, 9) List(376858, 0671027360, Book, 31, 1552, Angels & Demons) List(0671027360, 0385504209, 14178) List(296, 0385504209, Book, 19, 3049, The Da Vinci Code) List(0385504209, 0671027387, 10) List(424705, 0671027387, Book, 84, 429, Deception Point) List(0671027387, 0385504209, 15445) List(448, 0312966091, Book, 607, 141, Three To Get Deadly : A Stephanie Plum Novel (A Stephanie Plum Novel)) List(0312966091, 0312990456, 12) List(471848, 0312990456, Book, 304, 414, One for the Money (A Stephanie Plum Novel)) List(0312990456, 0312966091, 16904) List(448, 0312966091, Book, 607, 141, Three To Get Deadly : A Stephanie Plum Novel (A Stephanie Plum Novel)) List(0312966091, 0312971346, 13) List(104366, 0312971346, Book, 1401, 264, High Five (A Stephanie Plum Novel)) List(0312971346, 0312966091, 4164) List(448, 0312966091, Book, 607, 141, Three To Get Deadly : A Stephanie Plum Novel (A Stephanie Plum Novel)) List(0312966091, 0312976275, 14) List(475036, 0312976275, Book, 787, 324, Hot Six : A Stephanie Plum Novel (A Stephanie Plum Novel)) List(0312976275, 0312966091, 16978) List(448, 0312966091, Book, 607, 141, Three To Get Deadly : A Stephanie Plum Novel (A Stephanie Plum Novel)) List(0312966091, 0312980140, 15) List(158327, 0312980140, Book, 1084, 274, Seven Up (A Stephanie Plum Novel)) List(0312980140, 0312966091, 6266) List(457, B0000296JB, Music, 2439, 545, Make Yourself) List(B0000296JB, B00005QG9J, 17) List(119794, B00005QG9J, Music, 2488, 622, Morning View) List(B00005QG9J, B0000296JB, 4782) List(480, B0000296J9, Music, 1310, 42, Gunfighter Ballads & Trail Songs) List(B0000296J9, B0000026AG, 18) List(219235, B0000026AG, Music, 2015, 26, Johnny Horton - Greatest Hits) List(B0000026AG, B0000296J9, 8681) List(480, B0000296J9, Music, 1310, 42, Gunfighter Ballads & Trail Songs) List(B0000296J9, B0000024YO, 19) List(438745, B0000024YO, Music, 2293, 4, Marty Robbins - All-Time Greatest Hits) List(B0000024YO, B0000296J9, 15892) List(486, B0000296J3, Music, 1060, 46, Stardust) List(B0000296J3, B00004U2G7, 20) List(320882, B00004U2G7, Music, 709, 49, Red Headed Stranger) List(B00004U2G7, B0000296J3, 12192) List(486, B0000296J3, Music, 1060, 46, Stardust) List(B0000296J3, B00008BXK3, 21) List(394855, B00008BXK3, Music, 496, 15, The Essential Willie Nelson) List(B00008BXK3, B0000296J3, 14603) List(577, 0451163966, Book, 758, 316, One Flew Over the Cuckoo's Nest) List(0451163966, 0393312836, 28) List(196284, 0393312836, Book, 1073, 534, A Clockwork Orange (Norton Paperback Fiction)) List(0393312836, 0451163966, 7889) List(600, B000002V0T, Music, 3430, 55, American Patriot) List(B000002V0T, B00005QJDD, 30) List(162660, B00005QJDD, Music, 3422, 59, God Bless America) List(B00005QJDD, B000002V0T, 6409) List(622, B000002V07, Music, 1659, 79, So Tonight That I Might See) List(B000002V07, B000002TMY, 31) List(107745, B000002TMY, Music, 2588, 24, Among My Swan) List(B000002TMY, B000002V07, 4295) List(622, B000002V07, Music, 1659, 79, So Tonight That I Might See) List(B000002V07, B000003TA7, 32) List(269022, B000003TA7, Music, 4920, 55, Blind) List(B000003TA7, B000002V07, 10456) List(656, 0784017859, Video, 1250, 19, Denise Austin - Hit the Spot:Totally Firm) List(0784017859, 0784017786, 33) List(75908, 0784017786, Video, 794, 29, Denise Austin - Hit the Spot: Fat Burning Blast) List(0784017786, 0784017859, 3059) List(686, B00004RFIF, DVD, 4312, 51, Mystery Science Theater 3000 - The Brain That Wouldn't Die) List(B00004RFIF, B0000

how many product-pairs reference each other? can we just calculate it like this?

In [45]:
ProductsWithBalancedSimilarity.count()

Out[ 37 ]: 11202

does this mean, we have 11202 product-pairs that reference each other? Let's test if our data frame of mutually referencing products lists the pairs once or twice by looking at an example:

In [47]:
display(ProductsWithBalancedSimilarity.filter("(v1.id = '0790747324' and v2.id = 'B00007JMD8') or (v2.id = '0790747324' and v1.id = 'B00007JMD8')"))

v1 e1 v2 e2 List(21, 0790747324, DVD, 795, 140, The Time Machine) List(0790747324, B00007JMD8, 0) List(114681, B00007JMD8, DVD, 1076, 104, Journey to the Center of the Earth) List(B00007JMD8, 0790747324, 4581) List(114681, B00007JMD8, DVD, 1076, 104, Journey to the Center of the Earth) List(B00007JMD8, 0790747324, 4581) List(21, 0790747324, DVD, 795, 140, The Time Machine) List(0790747324, B00007JMD8, 0)

So the pairs are listed twice, which kind of makes sense, seeing how we constructed the data frame. This means there are 11202/2 = 5601 product-pairs that reference each other.

Now we want to explore if the "strongly connected components" of our graph are evenly distributed across all groups, which would mean that the probability of a DVD referencing to another DVD is the same as it referencing to e.g. a book, or if the groups tend to stick to themselves.

Here we first calculate the strongly connected components
(small warning: this command takes quite a while, ~1.17 hours, 238 jobs):

In [50]:
connectedComp = ProductGraph.stronglyConnectedComponents(maxIter=3)
#connectedComp.select("id", "component").groupBy("component").count().sort(desc("count")).show()

#ProductGraph.connectedComponents()
display(connectedComp)

The connected components are marked by the same value in the column "component". So if we group the data frame by "component" and "Group" and use the function "count()", we will have as many rows per "component" value as the number of different groups belonging to this cluster.

In [52]:
countGroupComponent= connectedComp.groupBy("component", "Group").count().sort("component")
display(countGroupComponent)

Now how many groups are taking part in each cluster?

In [54]:
countComponentsNumberOfGroups = countGroupComponent.groupBy("component").count().sort(desc("count"))
display(countComponentsNumberOfGroups)

Next we get an overview how many clusters there are with members of *n* groups:

In [56]:
numGroupsCount = countComponentsNumberOfGroups.withColumnRenamed("count", "NumberOfGroupsForComponent").groupBy("NumberOfGroupsForComponent").count()
display(numGroupsCount)

So how many clusters are there in total? Just adding up the values in column "count" in the last chunk results in the number of clusters:

In [58]:
display(numGroupsCount.agg({"count": "sum"}))

6751 is a very high number, seeing that there are just 10611 products in our dataset in total.

And what is the size distribution of the clusters in general (independent which groups its members belong to)?

In [61]:
countConnected = connectedComp.groupBy("component").count().sort(desc("count"))
display(countConnected)

So there is only one really huge cluster of strongly connected components with almost 1000 members. The next biggest one has only a size of 29.

How many of the 6751 clusters have more than 3 members?

In [64]:
countConnected.filter("count > 3").count()